glommio/src/channels/channel_mesh.rs

```diff
@@ -0,0 +1,312 @@
+            Self::Connected(receiver) => receiver.recv().await,
+        }
+    }
+}
+
+enum ReceiverProvider<T: Send + Copy> {
+    Local(LocalReceiver<T>),
+    Shared(SharedReceiver<T>),
+    Taken,
+}
+
+impl<T: Send + Copy> Default for ReceiverProvider<T> {
+    fn default() -> Self {
+        Self::Taken
+    }
+}
+
+impl<T: Send + Copy + 'static> ReceiverProvider<T> {
+    async fn get(self) -> Receiver<T> {
+        match self {
+            Self::Local(receiver) => Receiver::Local(receiver),
+            Self::Shared(receiver) => Receiver::Connected(receiver.connect().await),
+            Self::Taken => unreachable!(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Senders<T: Send + Copy> {
+    id: usize,
+    senders: Vec<Sender<T>>,
+}
+
+impl<T: Send + Copy> Senders<T> {
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    pub fn nr_peers(&self) -> usize {
+        self.senders.len()
+    }
+
+    pub async fn send_to(&self, peer_id: usize, msg: T) -> Result<(), T> {
+        self.senders[peer_id].send(msg).await
+    }
+
+    pub fn try_send_to(&self, peer_id: usize, msg: T) -> Result<(), T> {
+        self.senders[peer_id].try_send(msg)
+    }
+}
+
+#[derive(Debug)]
+struct Receivers<T: Send + Copy> {
+    id: usize,
+    receivers: Vec<Receiver<T>>,
+}
+
+impl<T: Send + Copy> Receivers<T> {
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    pub fn nr_peers(&self) -> usize {
+        self.receivers.len()
+    }
+
+    pub async fn recv_from(&self, peer_id: usize) -> Option<T> {
+        self.receivers[peer_id].recv().await
+    }
+}
+
+struct Notifier {
+    executor_id: usize,
+    sender: Option<SharedSender<usize>>,
+}
+
+impl Notifier {
+    fn new(sender: Option<SharedSender<usize>>) -> Self {
+        Self {
+            executor_id: Local::id(),
+            sender,
+        }
+    }
+}
+
+type ChannelProvider<T> = (Cell<SenderProvider<T>>, Cell<ReceiverProvider<T>>);
+
+type ChannelProviders<T> = Vec<Vec<ChannelProvider<T>>>;
+
+#[derive(Clone)]
+pub struct MeshBuilder<T: Send + Copy> {
+    nr_peers: usize,
+    notifiers: Arc<RwLock<Vec<Notifier>>>,
+    providers: Arc<Vec<Vec<ChannelProvider<T>>>>,
+}
+
+unsafe impl<T: Send + Copy> Send for MeshBuilder<T> {}
+
+impl<T: Send + Copy> Debug for MeshBuilder<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_fmt(format_args!(
+            "MeshBuilder {{ nr_peers: {} }}",
+            self.nr_peers
+        ))
+    }
+}
+
+impl<T: 'static + Send + Copy> MeshBuilder<T> {
+    pub fn new(nr_peers: usize, channel_size: usize) -> Self {
+        MeshBuilder {
+            nr_peers,
+            notifiers: Arc::new(RwLock::new(Vec::new())),
+            providers: Arc::new(Self::build_providers(nr_peers, channel_size)),
+        }
+    }
+
+    fn build_providers(nr_peers: usize, channel_size: usize) -> ChannelProviders<T> {
+        (0..nr_peers)
+            .map(|i| {
+                (0..nr_peers)
```

Comment:
does this mean we are also constructing a channel to ourselves? nr_peers x nr_peers.

Should probably exclude the diagonal of the matrix.